tox_conda/plugin.py

```diff
@@ -1,19 +1,187 @@
+import json
 import logging
+import os
+import shutil
+import sys
+from warnings import warn
 
 import pluggy
 
 hookimpl = pluggy.HookimplMarker("tox")
-log = logging.getLogger('conda')
+log = logging.getLogger("conda")
+
+
+on_win = bool(sys.platform == "win32")
 
 
 @hookimpl
 def tox_addoption(parser):
-    """Add a command line option for later use"""
-    parser.add_argument(
-        '--my-opt', action='store', help='my custom option')
+    parser.add_testenv_attribute(
+        name="conda_channels",
+        type="line-list",
+        default=("defaults"),
+        help="Conda channels for environment creation.",
+    )
+
+
+@hookimpl
+def tox_get_python_executable(envconfig):
+    """Return a python executable for the given python base name."""
+    if on_win:
+        return os.path.join(envconfig.envdir, "python.exe")
+    else:
+        return os.path.join(envconfig.envdir, "python")
 
 
 @hookimpl
 def tox_configure(config):
-    """Access your option during configuration"""
-    log.info("my option is: '%s'", config.option.my_opt)
+    """Called after command line options are parsed and ini-file has been read."""
+
+    # Directory to store conda files.
+    config.conda_workdir = os.path.join(config.toxworkdir, "conda")
+
+    # Override getting python executable.
+    envconfig_class = type(config.envconfigs[config.envlist[0]])
+    envconfig_class.get_envpython = get_conda_python
```

Comment:
I've tried to override the getting of the Python executable quite rudely here. 

However, it seems ``tox`` itself is often using the ``envconfig.envbindir`` to search for its python command even though it already stores it in two properties. Previous to this override, the ``envconfig.python_info.executable`` already used to point to the right one, only ``envconfig.envpython`` wasn't (hence the override).

Thinking this would fix things, I was wrong. 

In other occasions, ``tox`` does not refer to its explicit python properties already in the envclass, but instead starts looking for the "command" python instead, right in the environment's ``envconfig.envbindir``. 
https://github.com/tox-dev/tox/blob/f5707949159ffee516bf2e01a11d438f7ad49434/tox/venv.py#L118-L130

Conda puts the python executable right at the root, causing it not to be found and instead finds my system python interpreter.

This could be fixed in four ways:
* Reorder environment variables further up in the chain, causing it to find the local env's one instead of global python. (seems to be what @kalefranz was trying to do)
* Change ``tox`` to always use ``envconfig.envpython`` when it needs ``python`` and not start a search on path. The only real problem here is when it tries searching for python in the wrong dir as of now. The rest right in ``envconfig.envdir`` need not be included as far as I know. 
* Change ``tox`` to look in other directories, too. 
* More overrides in plugin.py (although I'd reather already remove this one and replace it by something more robust).

They're roughly sorted by preference. 

Is there a good reason that in ``tox.venv``, it starts searching for the ``python`` command as opposed to using the one from config? Moreover, why is there duplication in the config variables? Maybe there's some logic I'm failing to see.

Reply:
I would say definitely option 1. The reason why we search dynamically is that envpython is a Python name that needs to be resolved; however, envs are always run by using the resolved Python. E.g. user may say he wants to run with ``python3``; which might e.g. resolve to ``/home/magic/bin/python``.

Reply:
Hi, I'm new here, and so I could definitely be wrong, but it seems to me like `envpython` is actually a path to a concrete `python` executable within the test environment. In this case, wouldn't it make sense for `tox` to always use `envpython` when it uses python instead of searching on the path? This would make the implementation of this plugin *much* cleaner.